### PR TITLE
Adding new-SDKs

### PR DESCRIPTION
**Pull request by sightseer**
These are two methods of multipart uploading client.
File chunk method is to read each part of uploading file.
Uploading client method is to upload the whole file.

### CODE REVIEW

**Comment by sightseer** [comment]
Updated 20180825,2110

File: client/upload_client.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package client
+
+import (
```
Comment (Xuanwo) [inline client/upload_client.go]: import 要 follow 项目的规则，比如 request 中：

```go
import (
	"errors"
	"fmt"
	"net/http"
	"reflect"
	"strconv"
	"time"

	"github.com/pengsrc/go-shared/convert"

	"github.com/yunify/qingstor-sdk-go/logger"
	"github.com/yunify/qingstor-sdk-go/request/builder"
	"github.com/yunify/qingstor-sdk-go/request/data"
	"github.com/yunify/qingstor-sdk-go/request/signer"
	"github.com/yunify/qingstor-sdk-go/request/unpacker"
)
```

File: client/upload_client.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	"bytes"
+	qs "github.com/yunify/qingstor-sdk-go/service"
+	"os"
+)
+
+const smallestPartSize = 1024 * 1024 * 4
+
+//UploadClient uploads multiparts of large object
+func UploadClient(fd *os.File, objectKey string, partSize int, bucket *qs.Bucket) {
```
Comment (Xuanwo) [inline client/upload_client.go]: 为什么是一个 func ？应该实现一个 struct ，然后实现这个 struct 的一个方法。

File: client/file_chunk.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package client
+
+import (
+	"os"
+)
+
+const segmentSize = 1024
+
+//ReadObjectPart reads a part from an object and returns it
+func ReadObjectPart(fd *os.File, partSize int) []byte {
```
Comment (Xuanwo) [inline client/file_chunk.go]: 先实现一个 struct，然后实现这个 struct 的方法，不要直接写一个 func

File: client/upload_client.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+	_, completeErr := bucket.CompleteMultipartUpload(
+		objectKey,
+		&qs.CompleteMultipartUploadInput{
+			UploadID:    uploadID,
+			ObjectParts: partNumbers,
+		},
+	)
+	if completeErr != nil {
+		panic(completeErr)
```
Comment (Xuanwo) [inline client/upload_client.go]: 要遵循项目的规范，go sdk 中出错不会直接 panic ，然后将 err return，交由上级逻辑来处理。可以参考： https://github.com/yunify/qingstor-sdk-go/blob/master/request/request.go#L104

File: client/upload_client.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+const smallestPartSize = 1024 * 1024 * 4
+
+//UploadClient uploads multiparts of large object
+func UploadClient(fd *os.File, objectKey string, partSize int, bucket *qs.Bucket) {
+	if partSize < smallestPartSize {
+		panic("The part size is too small!")
+	}
+	output, initErr := bucket.InitiateMultipartUpload(
```
Comment (Xuanwo) [inline client/upload_client.go]: 最好不要在一个函数中塞进去太多复杂的逻辑，考虑一下拆分

File: client/upload_client.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		objectKey,
+		&qs.InitiateMultipartUploadInput{},
+	)
+	if initErr != nil {
+		panic(initErr)
```
Comment (Xuanwo) [inline client/upload_client.go]: 不要直接 panic，把 error return 会更好。

File: client/file_chunk.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	var chunk []byte
+	for {
+		n, _ = fd.Read(buf)
+		if n == 0 || lenBuf == partSize {
+			break
+		}
+		lenBuf = lenBuf + n
+		chunk = append(chunk, buf...)
+	}
+	return chunk[:lenBuf]
```
Comment (Xuanwo) [inline client/file_chunk.go]: 现在的实现会在内存中出现多次 copy，可以考虑参考一下 aws uploader 的实现： https://github.com/aws/aws-sdk-go/blob/master/service/s3/s3manager/upload.go#L507

File: client/upload_client.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+		_, uploadErr := bucket.UploadMultipart(
+			objectKey,
+			&qs.UploadMultipartInput{
+				UploadID:   uploadID,
+				PartNumber: &partCnt,
+				Body:       partBody,
+			},
+		)
+		if uploadErr != nil {
+			panic(uploadErr)
```
Comment (Xuanwo) [inline client/upload_client.go]: 不要直接 panic

**Review by Xuanwo — CHANGES_REQUESTED** [review]

**Comment by sightseer** [comment]
Updated 20170828,1210

File: client/upload_client.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+import (
+	"errors"
+	"io"
+	"os"
+
+	qs "github.com/yunify/qingstor-sdk-go/service"
+)
+
+type Uploader struct {
```
Comment (Xuanwo) [inline client/upload_client.go]: Follow the style like https://github.com/yunify/qingstor-sdk-go/blob/master/client/image/image.go#L99 , Upload is enough

File: client/upload_client.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+	Ubucket  *qs.Bucket
```
Comment (Xuanwo) [inline client/upload_client.go]: We will not expose the bucket, so just keep them private, ref: https://github.com/yunify/qingstor-sdk-go/blob/master/client/image/image.go#L99-L102

File: client/upload_client.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
+package client
+
+import (
+	"errors"
+	"io"
+	"os"
+
+	qs "github.com/yunify/qingstor-sdk-go/service"
```
Comment (Xuanwo) [inline client/upload_client.go]: service is unique in this package, we don't need to rename it to `qs`

File: client/upload_client.go [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+type Uploader struct {
+	Ubucket  *qs.Bucket
+	PartSize int
+}
+
+const smallestPartSize int = 1024 * 1024 * 4
+
+//UploadClient uploads multiparts of large object
+func (u *Uploader) UploadClient(fd *os.File, objectKey string) error {
```
Comment (Xuanwo) [inline client/upload_client.go]: In go, we use New or Init for create a new instance.

File: client/file_chunk.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
+package client
```
Comment (Xuanwo) [inline client/file_chunk.go]: use `upload` package will be better.

File: client/upload_client.go [hunk 14]
```diff
@@ -0,0 +1,93 @@
+package client
```
Comment (Xuanwo) [inline client/upload_client.go]: Put this file inot `client/upload` folder and use `upload` package.

File: client/upload_client.go [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+	return uploadID, nil
+}
+
+func (u *Uploader) UploadMultipart(fd *os.File, uploadID *string, objectKey string) ([]*qs.ObjectPartType, error) {
+	var partCnt int
+	partNumbers := []*qs.ObjectPartType{}
+	var objectReader ObjectReader
+
+	//var err error
+	objectReader.objectSize, _ = fd.Seek(0, io.SeekEnd)
```
Comment (Xuanwo) [inline client/upload_client.go]: With `NewSectionReader`, we don't need `objectSize`

**Review by Xuanwo — CHANGES_REQUESTED** [review]

**Comment by sightseer** [comment]
Updated 20170829,1649

**Comment by sightseer** [comment]
Add log, typeAssert method
Updated 20170829,2155

**Comment by sightseer** [comment]
Updated 20170830,1352

**Comment by sightseer** [comment]
Add file_chunk test file

**Comment by pengsrc** [comment]
What's the progress of this pull request?

**Comment by sightseer** [comment]
Add upload client test file

**Comment by sightseer** [comment]
updated 20170925,1810

**Comment by sightseer** [comment]
updated 20170925,2038

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**